### PR TITLE
Dominator deadcode problem fix

### LINKED ISSUE #1973: [Bug]: Dominator-deadcode problem
### Describe the issue:

The i++ node has no father though it's not  an ENTRY_POINT, so the only dominator of the node i++ is i++ itself. So when slither wants to find dominators for IF_LOOP node, it intersects the dominators of BEGIN_LOOP and i++, and gets nothing. Thus IF_LOOP has no immediate dominator which throws an assertion error

### Code example to reproduce the issue:

```solidity
contract C {
    function f() public {
        for (uint256 i = 1; i < 10; i++) {
            break;  // continue
        }
    }
}
```

### Version:

0.9.3

### Relevant log output:

```shell
'solc --version' running
'solc a.sol --combined-json abi,ast,bin,bin-runtime,srcmap,srcmap-runtime,userdoc,devdoc,hashes,compact-format --allow-paths . a.sol' running
Compilation warnings/errors on a.sol:
Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
--> a.sol

Warning: Source file does not specify required compiler version! Consider adding "pragma solidity ^0.7.5;"
--> a.sol

Warning: Unreachable code.
 --> a.sol:3:37:
  |
3 |         for (uint256 i = 1; i < 10; i++) {
  |                                     ^^^

Warning: Function state mutability can be restricted to pure
 --> a.sol:2:5:
  |
2 |     function f() public {
  |     ^ (Relevant source part starts here and spans across multiple lines).


ERROR:SlitherSolcParsing:
Failed to convert IR to SSA for C contract. Please open an issue https://github.com/crytic/slither/issues.
 
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 810, in main_impl
    ) = process_all(filename, args, detector_classes, printer_classes)
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 101, in process_all
    ) = process_single(compilation, args, detector_classes, printer_classes)
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 79, in process_single
    slither = Slither(target, ast_format=ast, **vars(args))
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 135, in __init__
    self._init_parsing_and_analyses(kwargs.get("skip_analyze", False))
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 155, in _init_parsing_and_analyses
    raise e
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 151, in _init_parsing_and_analyses
    parser.analyze_contracts()
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 541, in analyze_contracts
    self._convert_to_slithir()
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 774, in _convert_to_slithir
    raise e
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 769, in _convert_to_slithir
    contract.convert_expression_to_slithir_ssa()
  File "/usr/lib/python3.10/site-packages/slither/core/declarations/contract.py", line 1502, in convert_expression_to_slithir_ssa
    func.generate_slithir_ssa(all_ssa_state_variables_instances)
  File "/usr/lib/python3.10/site-packages/slither/core/declarations/function_contract.py", line 134, in generate_slithir_ssa
    compute_dominance_frontier(self.nodes)
  File "/usr/lib/python3.10/site-packages/slither/core/dominators/utils.py", line 98, in compute_dominance_frontier
    assert runner.immediate_dominator
AssertionError
ERROR:root:Error in a.sol
ERROR:root:Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 810, in main_impl
    ) = process_all(filename, args, detector_classes, printer_classes)
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 101, in process_all
    ) = process_single(compilation, args, detector_classes, printer_classes)
  File "/usr/lib/python3.10/site-packages/slither/__main__.py", line 79, in process_single
    slither = Slither(target, ast_format=ast, **vars(args))
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 135, in __init__
    self._init_parsing_and_analyses(kwargs.get("skip_analyze", False))
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 155, in _init_parsing_and_analyses
    raise e
  File "/usr/lib/python3.10/site-packages/slither/slither.py", line 151, in _init_parsing_and_analyses
    parser.analyze_contracts()
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 541, in analyze_contracts
    self._convert_to_slithir()
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 774, in _convert_to_slithir
    raise e
  File "/usr/lib/python3.10/site-packages/slither/solc_parsing/slither_compilation_unit_solc.py", line 769, in _convert_to_slithir
    contract.convert_expression_to_slithir_ssa()
  File "/usr/lib/python3.10/site-packages/slither/core/declarations/contract.py", line 1502, in convert_expression_to_slithir_ssa
    func.generate_slithir_ssa(all_ssa_state_variables_instances)
  File "/usr/lib/python3.10/site-packages/slither/core/declarations/function_contract.py", line 134, in generate_slithir_ssa
    compute_dominance_frontier(self.nodes)
  File "/usr/lib/python3.10/site-packages/slither/core/dominators/utils.py", line 98, in compute_dominance_frontier
    assert runner.immediate_dominator
AssertionError
```
```

### PR DESCRIPTION
**Pull request by Tiko7454**
Fixes #1973 the way I mentioned in the bug report

### CODE REVIEW

File: slither/solc_parsing/declarations/function.py [hunk 1]
```diff
@@ -1100,6 +1103,13 @@ def _parse_unchecked_block(self, block: Dict, node: NodeSolc, scope):
             node = self._parse_statement(statement, node, new_scope)
         return node
 
+    def _update_reachability(self, node: Node):
```
Comment (smonicas) [inline slither/solc_parsing/declarations/function.py:1106]: Can you add `-> None` as type hint

**Review by smonicas — CHANGES_REQUESTED** [review]

File: slither/core/dominators/utils.py [hunk 2]
```diff
@@ -9,9 +9,16 @@
...
         return set()
-    ret = node.fathers[0].dominators
-    for pred in node.fathers[1:]:
-        ret = ret.intersection(pred.dominators)
+    if not any(father.is_reachable for father in node.fathers):
+        return set()
+
+    ret = set()
+    for pred in node.fathers:
+        ret = ret.union(pred.dominators)
```
Comment (montyly) [inline slither/core/dominators/utils.py:17]: Mhh I am confused, can you explain why are we doing the union over the predecessor dominators, and then their interaction?

It seem to me that this loop should be removed - unless I am missing something here

Comment (Tiko7454) [reply slither/core/dominators/utils.py:17]: It's the safest way to start set intersections. I could store the first set as the base and start interacting all sets with that but the first set may not exist. So the safest way is to get every element from every set, and start to intersect every set with that big set

Comment (montyly) [reply slither/core/dominators/utils.py:17]: gotcha that makes sense :)

**Comment by montyly** [comment]
thanks @Tiko7454

**Comment by montyly** [comment]
For visibility, there was an issue with the recent add of the vyper support (see https://github.com/crytic/slither/issues/2171). We will investigate this further

**Pull request merged**